NetKet/Sampler/abstract_sampler.hpp

```diff
@@ -50,8 +51,35 @@ class AbstractSampler {
     this->Reset(true);
   }
 
+  virtual void SetMachineFunc(
+      std::function<double(const Complex&)> machine_func) {
+    machine_func_.Set(machine_func);
+  }
+
+  virtual std::function<double(const Complex&)> GetMachineFunc() {
+    return machine_func_.Get();
+  }
+
+  class MachineFunc {
+   private:
+    std::function<double(const Complex&)> mf_;
+
+   public:
+    MachineFunc() {
+      // By default, the L2-norm is used for sampling
+      mf_ = static_cast<double (*)(const Complex&)>(&std::norm);
+    }
+    double operator()(const Complex& c) { return mf_(c); }
+
+    void Set(std::function<double(const Complex&)> machine_func) {
+      mf_ = machine_func;
+    }
+    std::function<double(const Complex&)> Get() { return mf_; }
+  };
+
  protected:
   default_random_engine& GetRandomEngine() { return engine_.Get(); }
+  MachineFunc machine_func_;
```

Comment:
Random number generator `engine_` is private, but `machine_func_` is exposed to the children, why is that? I think `GetMachineFunction` should be used to obtain a reference to `machine_func_` similar to how `GetRandomEngine` works.

One of the lesser known facts about `std::function` is that it can have a empty state. This is undesirable in our case. By initialising `machine_func_` with `std::norm` and checking for the empty state in `SetMachineFunc` we can prevent ever having an empty state.

Reply:
Sure but I don't want that all the derived samplers need to initialize it on its default value, I would like that to be done in the abstract class somehow

Reply:
actually I would make `engine_` protected too, I don't see a valid reason for not just using it without using setters and getters, maybe I am missing something?

Reply:
I think @femtobit designed `engine_` to be private and not protected, can we change it to protected ?

Reply:
@gcarleo typically, the reason to use getters and setters rather than exposing a member variable is to impose pre- and post-conditions. So the question is whether we trust the implementation of derived classes. If we ever wish to allow extending NetKet through C++, we should probably go the safer (getter + setter) route. Otherwise, NetKet maintainers are the only people who'll ever touch `machine_func_` and we can make it protected.

Reply:
fine, this is really too much precaution, but anyway... it's done!